File: problems/00036-valid-sudoku/go/main.go
package main

// Runs a closure function with the map's values and returns true if
// any of the value results true with that func.
func mapAnyFunction[K comparable, V any](m map[K]V, f func(V) bool) bool {
	for _, v := range m {
		if f(v) {
			return true
		}
	}
	return false
}

func isValidSudoku(board [][]byte) bool {
	rowMap := make(map[byte]int, 9)
	collumnMap := make(map[byte]int, 9)
	var curDigit byte
	for y := range 9 {
		for x := range 9 {
			curDigit = board[y][x]
			if curDigit != '.' {
				rowMap[curDigit]++
			}
			curDigit = board[x][y]
			if curDigit != '.' {
				collumnMap[curDigit]++
			}
		}
		if mapAnyFunction(rowMap, func(val int) bool {
			return val > 1
		}) || mapAnyFunction(collumnMap, func(val int) bool {
			return val > 1
		}) {
			return false
		}
		clear(rowMap)
		clear(collumnMap)
	}
	gridMap := make(map[byte]int, 9)
	for y := 1; y < 9; y += 3 {
		for x := 1; x < 9; x += 3 {
			for a := y - 1; a < y+2; a++ {
				for b := x - 1; b < x+2; b++ {
					curDigit = board[a][b]
					if curDigit == '.' {
						continue
					}
					gridMap[curDigit]++
				}
			}
			if mapAnyFunction(gridMap, func(val int) bool {
				return val > 1
			}) {
				return false
			}
			clear(gridMap)
		}
	}
	return true
}

File: problems/00036-valid-sudoku/rust/src/main.rs
#![allow(dead_code)]
struct Solution {}

use std::collections::HashMap;

impl Solution {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let mut row_map = HashMap::new();
        let mut col_map = HashMap::new();
        for a in 0..9 {
            for b in 0..9 {
                match board[a][b] {
                    '.' => {}
                    x => {
                        row_map.entry(x).and_modify(|val| *val += 1).or_insert(1);
                    }
                }
                match board[b][a] {
                    '.' => {}
                    x => {
                        col_map.entry(x).and_modify(|val| *val += 1).or_insert(1);
                    }
                }
            }
            if row_map.values().any(|val| val > &1) || col_map.values().any(|val| val > &1) {
                return false;
            }
            row_map.clear();
            col_map.clear();
        }
        let mut grid_map = HashMap::new();
        for a in (1..9).step_by(3) {
            for b in (1..9).step_by(3) {
                #[allow(clippy::needless_range_loop)]
                for i in (a - 1)..(a + 2) {
                    for j in (b - 1)..(b + 2) {
                        match board[i][j] {
                            '.' => {}
                            x => {
                                grid_map.entry(x).and_modify(|val| *val += 1).or_insert(1);
                            }
                        }
                    }
                }
                if grid_map.values().any(|val| val > &1) {
                    return false;
                }
                grid_map.clear();
            }
        }
        true
    }
}

fn main() {
    println!("This program doesn't have testing since there's no suitable way of complex testing");
}

